openmc/mesh.py

```diff
@@ -1137,6 +1137,70 @@ def from_hdf5(cls, group):
 
         return mesh
 
+    @classmethod
+    def from_domain(
+        cls,
+        domain,
+        dimension=[100, 100, 100],
+        mesh_id=None,
+        phi_grid=[0.0, 2*pi],
+        name=''
+    ):
+        """Create mesh from an existing openmc cell, region, universe or
+        geometry by making use of the objects bounding box property. phi_grid
+        is set to have a range of 0 to 2π by default.
+
+        Parameters
+        ----------
+        domain : {openmc.Cell, openmc.Region, openmc.Universe, openmc.Geometry}
+            The object passed in will be used as a template for this mesh. The
+            bounding box of the property of the object passed will be used to
+            set the r_grid, z_grid ranges.
+        dimension : Iterable of int
+            The number of equally spaced mesh cells in each direction (r_grid,
+            phi_grid, z_grid)
+        mesh_id : int
+            Unique identifier for the mesh
+        phi_grid : numpy.ndarray
+            1-D array of mesh boundary points along the phi-axis in radians.
+            The default value is [0, 2π], i.e. the full phi range.
+        name : str
+            Name of the mesh
+
+        Returns
+        -------
+        openmc.RegularMesh
+            RegularMesh instance
+
+        """
+        cv.check_type(
+            "domain",
+            domain,
+            (openmc.Cell, openmc.Region, openmc.Universe, openmc.Geometry),
+        )
+
+        mesh = cls(mesh_id, name)
+
+        # loaded once to avoid reading h5m file repeatedly
+        cached_bb = domain.bounding_box
+        max_bounding_box_radius = max(
+            [
+                cached_bb[0][0],
+                cached_bb[0][1],
+                cached_bb[1][0],
+                cached_bb[1][1],
+            ]
+        )
```

Comment:
This would only work when the domain is centered at (0, 0)... which reminds me that our cylindrical and spherical meshes don't work for things that are not centered :disappointed:

Reply:
@paulromano that's true. Maybe we could add a centre to these meshes. Shouldn't be too complicated.

Reply:
Created issue to track this idea https://github.com/openmc-dev/openmc/issues/2248